drivers/base/regmap/regmap-sdw-mbq.c

```diff
@@ -31,12 +32,40 @@ static int regmap_sdw_mbq_size(struct regmap_mbq_context *ctx, unsigned int reg)
 	return size;
 }
 
-static int regmap_sdw_mbq_write(void *context, unsigned int reg, unsigned int val)
+static int regmap_sdw_mbq_deferrable(struct regmap_mbq_context *ctx, unsigned int reg)
 {
-	struct regmap_mbq_context *ctx = context;
-	struct device *dev = ctx->dev;
-	struct sdw_slave *slave = dev_to_sdw_dev(dev);
-	int mbq_size = regmap_sdw_mbq_size(ctx, reg);
+	if (ctx->cfg.deferrable)
+		return ctx->cfg.deferrable(ctx->dev, reg);
+
+	return 0;
+}
+
+static int regmap_sdw_mbq_poll_busy(struct sdw_slave *slave, unsigned int reg,
+				    unsigned long timeout_us, unsigned long retry_us)
+{
+	struct device *dev = &slave->dev;
+	int val, ret;
+
+	dev_dbg(dev, "Deferring transaction for 0x%x\n", reg);
+
+	reg = SDW_SDCA_CTL(SDW_SDCA_CTL_FUNC(reg), 0, SDW_SDCA_FUNCTION_STATUS, 0);
+
+	ret = read_poll_timeout(sdw_read_no_pm, val,
+				val < 0 || !(val & SDW_SDCA_FUNCTION_BUSY),
+				timeout_us, retry_us, false, slave, reg);
```

Comment:
probably worthy of a comment saying that this cannot be a loop where COMMAND_IGNORED is returned again after polling on FUNCTION_BUSY.

But I am not actually sure that the spec guarantees this. And with two regmap implementations it's possible in theory that the function becomes busy again become something else is updated in the other regmap while we wait. This could be awfully racy if we don't have a sequential behavior with a single regmap.

Reply:
Yeah my primary aim with these changes is to avoid the situation where you have two regmaps because it makes interacting with ASoC a right mess. So I guess my take on this would be access to the SDCA registers is blocked because it all happens under the regmap lock, so you can't get a second defer. I can add a comment to that effect if we like.

Reply:
hm... yes looking through the spec I think the spec is perhaps slightly broken here. It does say that "a function may set the bit to 1 when a resource within the function means that it is not currently able to perform actions (eg. because an internal processor is initializing)" Which strongly implies that a function can asynchronously of its own accord set function busy, but I think we have to assume that means only init time, as per the example. Because it also states that "mipi-sdca-function-busy-max-delay reports the longest time that Function_Busy will remain at 1", but obviously that information is useless to the host if the device can asynchronously function busy, since your attempts to access it might just get unlucky and keep being between the periods its free. So the only reasonable wait would be forever.

Reply:
The spec also says "A Function in a Peripheral Device can work on a maximum of one deferred read at a time". so presumably we cannot end-up in a doom loop of Function_Busy being set/cleared multiple times by the Peripheral without any read/write command in between.

Reply:
That is good that it rules that one out, as I suspect that is really the only issue in practice. But the "a function may set the bit to 1 when a resource within the function means that it is not currently able to perform actions" still basically gives functions permission to busy themselves asynchronously from register transactions, which basically means there is no sensible wait time. But I will choose to believe that isn't really a thing for now until we are forced to confront it.